app/grandchallenge/algorithms/models.py

```diff
@@ -584,15 +663,68 @@ def get_remaining_complimentary_jobs(self, *, user):
             return 0
 
     def get_remaining_non_complimentary_jobs(self, *, user):
-        user_credits = Credit.objects.get(user=user).credits
-        spent_credits = Job.objects.credits_consumed_past_month(user=user)[
-            "total"
-        ]
-
-        credits_left = user_credits - spent_credits
+        try:
+            credits_left = self.get_remaining_specific_credits(
+                user=user, algorithm=self.algorithm
+            )
+        except ObjectDoesNotExist:
+            credits_left = self.get_remaining_general_credits(user=user)
 
         return max(credits_left, 0) // max(self.algorithm.credits_per_job, 1)
 
+    @staticmethod
+    def get_remaining_specific_credits(*, user, algorithm):
+        user_credit = AlgorithmUserCredit.objects.active_credits().get(
+            user=user,
+            algorithm=algorithm,
+        )
+
+        spent_credits = Job.objects.filter(
+            creator=user_credit.user,
+            is_complimentary=False,
+            created__date__gte=user_credit.valid_from,
+            created__date__lte=user_credit.valid_until,
+            algorithm_image__algorithm=user_credit.algorithm,
+        ).aggregate(
+            total=Sum("credits_consumed", default=0),
+        )
+
+        return user_credit.credits - spent_credits["total"]
+
+    @staticmethod
+    def get_remaining_general_credits(*, user):
+        if user.username == settings.ANONYMOUS_USER_NAME:
+            return 0
+
+        user_credits = settings.ALGORITHMS_GENERAL_CREDITS_PER_MONTH_PER_USER
+
+        user_algorithms_with_active_credits = (
+            AlgorithmUserCredit.objects.active_credits()
+            .filter(
+                user=user,
+            )
+            .select_related("algorithm")
```

Comment:
Not sure how much time it really saves, but I think it's be more efficient if you just returned the algorithm pk here: 
```suggestion
            .values_list("algorithm__pk", flat=True)
```

and then in the next query filter on those:
`.exclude(algorithm_image__algorithm__pk__in=user_algorithms_with_active_credits)`